src/rust-crypto/libolm_migration.ts

```diff
@@ -380,3 +382,91 @@ async function getAndDecryptCachedSecretKey(
+ * By default, if the legacy session didn't have the private MSK, the migrated session will revert to unverified,
+ * even if the user has verified the session in the past.
+ * This only occurs if the private MSK was not cached in the crypto store (USK and SSK private keys won't help
+ * to establish trust: the trust is rooted in the MSK).
+ * Rust crypto will only consider the current session as trusted if we import the private MSK itself.
+ *
+ * We could prompt the user to verify the session again, but it's probably better to just mark the user identity
+ * as locally verified if it was before.
+ *
+ * See https://github.com/element-hq/element-web/issues/27079
+ *
+ * @param args - Argument object.
+ */
+export async function migrateLegacyLocalTrustIfNeeded(args: {
+    /** The legacy crypto store that is migrated. */
+    legacyCryptoStore: CryptoStore;
+    /** The migrated rust crypto stack. */
+    rustCrypto: RustCrypto;
+    /** The logger to use */
+    logger: Logger;
+}): Promise<void> {
+    const { legacyCryptoStore, rustCrypto, logger } = args;
+    // Now get the cross-signing identity from rust.
+    // If this is null that means that there are no cross-signing keys published server side.
+    const rustOwnIdentity = await rustCrypto.getOwnIdentity();
+    if (!rustOwnIdentity) {
+        // There are no cross-signing keys published server side, so nothing to do here.
+        return;
+    }
+    if (rustOwnIdentity.isVerified()) {
+        // The rust session already trusts the keys, so again, nothing to do.
+        return;
+    }
+
+    // This would be null is the user never verified their identity in the legacy session.
+    const legacyLocallyTrustedMSK = await getLegacyTrustedPublicMasterKeyBase64(legacyCryptoStore);
+    if (!legacyLocallyTrustedMSK) {
+        // Nothing to do, the legacy session was not verified
+        return;
+    }
+
+    const mskInfo: CrossSigningKeyInfo = JSON.parse(rustOwnIdentity.masterKey);
+    if (!mskInfo.keys || Object.keys(mskInfo.keys).length === 0) {
+        // This should not happen, but let's be safe
+        logger.error("Post Migration | Unexpected error: no master key in the rust session.");
+        return;
+    }
+    const rustSeenMSK = Object.values(mskInfo.keys)[0];
+
+    if (rustSeenMSK && rustSeenMSK == legacyLocallyTrustedMSK) {
+        logger.info(`Post Migration: Migrating legacy trusted MSK: ${legacyLocallyTrustedMSK} to locally verified.`);
+        // Let's mark the user identity as locally verified as part of the migration.
+        await rustOwnIdentity!.verify();
+        // This will also sign the MSK with the current device key, and return the request to be sent to the server.
+        // We ignored it in our case. As the local trust was migrated and not actually done by the user it looks
+        // sane to not publish the signature. Anyhow, these device's signatures are not considered by the rust-sdk to
+        // establish the trust of the user identity.
```

Comment:
I think it's helpful to record that `.verify` doesn't persistently store the signature, and if we drop the request here it won't be retried.

Putting all the above together, does this seem accurate?

```suggestion
        // As well as marking the MSK as trusted, `OlmMachine.verify` returns a
        // `SignatureUploadRequest` which will publish a signature of the MSK using
        // this device. In this case, we ignore the request: since the user hasn't
        // actually re-verified the MSK, we don't publish a new signature. (`.verify`
        // doesn't store the signature, and if we drop the request here it won't be
        // retried.)
        //
        // Not publishing the signature is consistent with the behaviour of
        // matrix-crypto-sdk when the private key is imported via
        // `importCrossSigningKeys`, and when the identity is verified via interactive
        // verification.
        //
        // [Aside: device signatures on the MSK are not considered by the rust-sdk to
        // establish the trust of the user identity so in any case, what we actually do
        // here is somewhat moot.]
```